pandas/core/frame.py

```diff
@@ -528,9 +528,14 @@ def __init__(
                 values = cast_scalar_to_array(
                     (len(index), len(columns)), data, dtype=dtype
                 )
-                mgr = init_ndarray(
-                    values, index, columns, dtype=values.dtype, copy=False
-                )
+                if isinstance(values, list):
```

Comment:
this is a crazy api here. Why are you doing this?

Reply:
I try to clarify this in the comments, but I can understand why this could be confusing. There are two distinct possibilities:

1. `data` is a non-extension type, in which case `cast_scalar_to_array` will return a numpy ndarray
2. `data` is an extension type, in which cast `cast_scalar_to_array` will return a list of extension arrays

The two cases also require two separate instantiations of the manager, since they are different data types.

I'm not sure we can do better here. If `data` is an extension type, then it can't be put into a Numpy array and if `data` is not an extension type, we don't want to create a list of separate arrays, we want to create a single higher-dimensional array

Reply:
Ideally the flow is something like

```python
# in DataFrame.__init__
if values_is_a_scalar:  # the case we're in here
    # discover if we're an extension dtype (make a list of arrays)
    # or a numpy dtype (make a single ndarray)
    dtype = dtype or infer_dtype(values, pandas_dtype=True)  # infer if needed
    if is_extension_array_dtype(dtype):
        values = cast_scalar_to_list_of_extension_arrays(...)
        mgr = arrays_to_mgr(...)
    else:
        values = cast_scalar_to_array(...)
        mgr = init_ndarray(...)
```


So the only extra complexity in `DataFrame.__init__` reflects the *fundamental* fact that an "ndarray" scalar has to take a different path from the "extension" scalar.

However, I think that right now `dtype` might be `None`, since `cast_array_to_scalar` does a bit of inference. So ideally `cast_array_to_scalar` only does the array construction, and doesn't do the dtype inference. Does that make sense @justinessert?

Reply:
@TomAugspurger Happy to follow all or some of that method. I also just pushed a way to flatten out the `else` case, curious if you want to keep any element of this?
```
# DOESN'T WORK FOR TIMEZONE, PLEASE SUGGEST BETTER WAY TO DO THIS CHECK
        elif isinstance(data, (Period, Interval)):
            values = cast_scalar_to_array((len(index), len(columns)), data, dtype=dtype)

            if index is not None and columns is not None:
                mgr = arrays_to_mgr(values, columns, index, columns, dtype=None)
            else:
                raise ValueError("DataFrame constructor not properly called!")
        else: # Else case Identical to the way it was previously!
            try: ...
            except (ValueError, TypeError) as err: ...

            if arr.ndim == 0 and index is not None and columns is not None:
                values = cast_scalar_to_array(
                    (len(index), len(columns)), data, dtype=dtype
                )
                mgr = init_ndarray(
                    values, index, columns, dtype=values.dtype, copy=False
                )
            else:
                raise ValueError("DataFrame constructor not properly called!")
```

Reply:
Hmm hard to say, but I don't think we want to be treating Period / Interval specially. Ideally we're just looking at the dtype implied by the scalar and using that to determine which path we take.